### ISSUE TITLE
Porting to webassembly

### ISSUE THREAD

**Issue by saolsen** [jart/blink#8]
Hey,

I'm thinking about trying to get blink running in the browser (via webassembly). My goal isn't just to run c code in the browser (could just use emscripten for that) but to actually run ape x64 executables in an interpreter where I can build a debugger and some visualization tools to see what's happening. (Similar to what blink already does).
I'm wondering if anybody has tried to compile blink for webassembly or what you think some of the challenges would be.

**Comment by jart** [comment]
Sounds awesome. I haven't tried it myself. I imagine the only major obstacle might be 32-bit. JavaScript only supports 32-bit integers. Blink runs on i386 and other 32-bit CPUs, and is regularly tested on them. However Blink was created in a 64-bit world, and as such, Blink has a 64-bit bias, therefore, you might not get optimal performance out of Blink in your 32-bit environments.

**Comment by unicomp21** [comment]
https://github.com/WebAssembly/tail-call/issues/15#issuecomment-1357820841

We're still waiting on tail calls for webassembly, lol.  Wish there was another option.

**Comment by unicomp21** [comment]
Someday we'll be able to use co_await in webassembly, hope I'm still around to see it happen, lol.

https://github.com/WebAssembly/tail-call/issues/14#issue-1058389687

**Comment by unicomp21** [comment]
https://github.com/emscripten-core/emscripten/issues/10991#issuecomment-974226917

**Comment by jart** [comment]
Blink doesn't need tail calls.

Also, good news! Blink is now stable on 32-bit platforms. I don't know if webassembly is multi-core, but the threading issues Blink was having earlier on 32-bit have been resolved.

Your biggest obstacle is most likely going to be completely replacing everything in blink/syscall.c so it interfaces with WASI or something similar instead.

**Comment by unicomp21** [comment]
https://github.com/copy/v86

**Comment by unicomp21** [comment]
@gornishanov @tlively @madmongo1 @tomoliv any ideas on easiest way to implement linux syscall.c for webassembly?

https://github.com/jart/blink/issues/8#issuecomment-1370735621

https://github.com/WebKit/WebKit/pull/2065

**Comment by unicomp21** [comment]
Dumb/crazy question, what sort of perf hit if blink runs nested within itself?  I'm wondering if the outer vm could implement syscall.c and handle threading etc. using co_await?

**Comment by jart** [comment]
It's possible to run Blink within Blink. There's noticeable slowdown, but it's not a showstopper. There's caveats, such as you can only have a single of the nested Blink instances make use of the linear memory optimization. The other nestings need to pass `blink -m` to turn it off, otherwise the memory allocations will collide.

**Comment by Vogtinator** [comment]
> any ideas on easiest way to implement linux syscall.c for webassembly?

emscripten provides a surprisingly wide set of runtime APIs which might even be enough to "just work" already.

**Comment by Vogtinator** [comment]
> Also, good news! Blink is now stable on 32-bit platforms. I don't know if webassembly is multi-core, but the threading issues Blink was having earlier on 32-bit have been resolved.

With web workers and SharedArrayBuffers it's possibly to effectively have threads and emscripten even exposes those through pthreads as much as possible.

> > any ideas on easiest way to implement linux syscall.c for webassembly?
> 
> emscripten provides a surprisingly wide set of runtime APIs which might even be enough to "just work" already.

I gave it a quick try and it does almost build out-of-the-box with just `emmake make -j8`. It just complains about `sa_len` and `wait4` missing. With those worked around I get a `blink.wasm` of unknown quality, not tested.

**Comment by jart** [comment]
I got it to compile too. When I tried to run Blink in Node, this happened, and I have no idea what it means.

```
master jart@turfwar:~/blink$ node o//blink/blink
/home/jart/blink/o/blink/blink:2148
  function ___invoke_$struct_Machine*_$struct_System*_$struct_Machine*(
                                    ^

SyntaxError: Unexpected token '*'
    at wrapSafe (internal/modules/cjs/loader.js:915:16)
    at Module._compile (internal/modules/cjs/loader.js:963:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)
    at internal/main/run_main_module.js:17:47
```

I configured the Makefile to build an HTML file and ran it in the browser. I got the same thing.

![image](https://user-images.githubusercontent.com/49262/210613350-08abd58a-0931-47c5-a8b1-bdef2fa67caf.png)

I'm going to push the fixes I made right now. Could you please help me figure out what's wrong?

**Comment by Vogtinator** [comment]
You might need newer LLVM or emscripten: https://github.com/emscripten-core/emscripten/issues/12551

I'm using emscripten main with clang 15.0.6 (should be 16 mmeanwhile, but it works :shrug:)

**Comment by Vogtinator** [comment]
Works!

```
/tmp/blink> node o/blink/blink -m /cwd/third_party/cosmo/tinyhello.elf
hello world
```

The WASM page size is 64KiB, which means that https://github.com/jart/blink/issues/14 happens, but by doing the awful hack of just pretending that it's actually 4096 it can even run `busybox-static` from the host here:

```
/tmp/blink> (cd /usr/bin; node /tmp/blink/o/blink/blink -m /cwd/busybox-static sh -c "echo Hello world!")
I2023-01-04T19:25:47.197000:blink/syscall.c:2610:42 missing syscall 0x111
I2023-01-04T19:25:47.198000:blink/syscall.c:2610:42 missing syscall 0x14e
warning: unsupported syscall: __syscall_prlimit64

I2023-01-04T19:25:47.200000:blink/syscall.c:1857:42 getrandom() flags not supported yet
Hello world!
```

It did require some hacks and workarounds though:

* Expose the working directory in the virtual filesystem below `/cwd`
* emscripten does not pass `envp` to main, use `environ` instead (crashes in `LoadArgv` otherwise)

```diff
diff --git a/blink/blink.c b/blink/blink.c
index f6c0506..a241fd0 100644
--- a/blink/blink.c
+++ b/blink/blink.c
@@ -155,8 +155,10 @@ static void HandleSigs(void) {
   unassert(!sigaction(SIGSEGV, &sa, 0));
 #endif
 }
-
+#include <emscripten.h>
 int main(int argc, char *argv[], char **envp) {
+  EM_ASM({FS.mkdir('/cwd'); FS.mount(NODEFS, {root : '.'}, '/cwd');});
+  if(!envp) envp = environ;
   g_blink_path = argc > 0 ? argv[0] : 0;
   GetOpts(argc, argv);
   if (optind_ == argc) PrintUsage(argc, argv, 48, 2);
```

* Function pointers work differently, they don't really contain an address so comparing `> 4096` fails:

```diff
diff --git a/blink/debug.h b/blink/debug.h
index a996413..5e64aa9 100644
--- a/blink/debug.h
+++ b/blink/debug.h
@@ -10,7 +10,7 @@
 #define IB(x)                      \
   __extension__({                  \
     __typeof__(x) x_ = (x);        \
-    unassert((intptr_t)x_ > 4096); \
+    unassert((intptr_t)x_ > 0); \
     x_;                            \
   })
 #else
```

* emscripten's `mmap` does not support address hints:

```diff
diff --git a/blink/memorymalloc.c b/blink/memorymalloc.c
index 3d0113b..b0b9266 100644
--- a/blink/memorymalloc.c
+++ b/blink/memorymalloc.c
@@ -64,6 +64,10 @@ void FreeBig(void *p, size_t n) {
 
 void *AllocateBig(size_t n) {
   void *p;
+#ifdef __EMSCRIPTEN__
+  p = Mmap(NULL, n, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0, "big");
+  return p != MAP_FAILED ? p : 0;
+#endif
   u8 *brk;
   if (!(brk = atomic_load_explicit(&g_allocator.brk, memory_order_relaxed))) {
     // we're going to politely ask the kernel for addresses starting
```

I built it with `emmake make -j8 && emcc -g o//blink/blink.o o//blink/blink.a -lm -pthread -lrt -o o//blink/blink -s INITIAL_MEMORY=1073741824 -s EXIT_RUNTIME=1 -lnodefs.js`

**Comment by jart** [comment]
Wow. I'm still catching up. Quick question. Did you have any problems with `wait4`? My build is complaining about that being undefined.

**Comment by jart** [comment]
I made a bunch more changes and I'm now blocked on this error.

```
master jart@turfwar:~/blink$ node o//blink/blink
requested a shared WebAssembly.Memory but the returned buffer is not a SharedArrayBuffer, indicating that while the browser has SharedArrayBuffer it does not have WebAssembly threads support - you may need to set a flag
(on node you may need: --experimental-wasm-threads --experimental-wasm-bulk-memory and/or recent version)
/home/jart/blink/o/blink/blink:163
      throw ex;
      ^

Error: bad memory
    at Object.<anonymous> (/home/jart/blink/o/blink/blink:820:13)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)
    at internal/main/run_main_module.js:17:47
$?=7 master jart@turfwar:~/blink$ type node
node is hashed (/home/jart/vendor/emsdk/node/14.18.2_64bit/bin/node)
```

I got a little further in the browser. Not sure what to do next.

![image](https://user-images.githubusercontent.com/49262/210630435-8d02e798-8fc8-445a-93c5-d94a754872d1.png)

**Comment by Vogtinator** [comment]
> Wow. I'm still catching up. Quick question. Did you have any problems with wait4? My build is complaining about that being undefined.

Yeah, I had to stub that out. FWICT that's a bug in emscripten in some way, as it's available in the system headers and there's also a stub for `__syscall_wait4`

> I made a bunch more changes and I'm now blocked on this error.

Might just be the version of node, it might not support SharedArrayBuffer backed memory for WASM. I'm using v19.3.0 here.

> I got a little further in the browser. Not sure what to do next.

That's where the "fun" part starts: Somehow making use of `blink.js` in the web page. Here's a minimal PoC:

`pre.js`:
```javascript
function fileLoad(event, filename) {
    var file = event.target.files[0];
    var reader = new FileReader();
    reader.onloadend = function(event) {
      if(event.target.readyState == FileReader.DONE)
        FS.writeFile(filename, new Uint8Array(event.target.result), {encoding: 'binary'});
    };
    reader.readAsArrayBuffer(file);
}

let fileInput = document.createElement("input");
fileInput.setAttribute("type", "file");
fileInput.onchange = () => { fileLoad(event, "executable"); };
document.body.appendChild(fileInput);

let button = document.createElement("button");
button.innerText = "Start";
button.onclick = () => { Module.callMain(["executable"]); };
document.body.appendChild(button);
```

Build with `emcc o//blink/blink.o o//blink/blink.a  -lm -pthread -lrt -o o//blink/blink.html -s INVOKE_RUN=0 -s EXPORTED_RUNTIME_METHODS=callMain -s INITIAL_MEMORY=1073741824 -s EXIT_RUNTIME=1 --emrun --pre-js pre.js`

At some point the best option is probably to expose some kind of API to JS, depending on what the actual use cases are.

**Comment by unicomp21** [comment]
@Vogtinator @jart you guys are amazing.  This is great!

**Comment by 0xhiro** [comment]
Well done guys! Very Cool!

**Comment by Rucadi** [comment]
This is awesome, WILL hack with this :D

**Comment by pannous** [comment]
please someone host their compiled blink.wasm !

**Comment by 0xhiro** [comment]
> please someone host their compiled blink.wasm !

Hmmmmm........

**Comment by Vogtinator** [comment]
I pushed a github workflow for emscripten HTML builds: https://github.com/Vogtinator/blink/actions/workflows/emscripten.yml

To build for node instead, uncomment the NODEFS mounting in `blink/web.h` and build with `emcc -O2 o//blink/blink.o o//blink/blink.a -lm -pthread -lrt -o o//blink/blink -s INITIAL_MEMORY=1073741824 -s EXIT_RUNTIME=1 -lnodefs.js`.

**Comment by unicomp21** [comment]
@derekcollison I'm wondering if there could be implications/synergy here for nats?  ie tunneling tcp syscalls, etc.?

**Comment by derekcollison** [comment]
> @derekcollison I'm wondering if there could be implications/synergy here for nats? ie tunneling tcp syscalls, etc.?

How so? Maybe running a nats-server in the browser?

**Comment by unicomp21** [comment]
Yes, or the networking layer for many vm's running in browsers?

**Comment by jart** [comment]
Tweeted https://mobile.twitter.com/JustineTunney/status/1613895681038770182

@trungnt2910 @Vogtinator Would you both be interested in upstreaming your work? I've just added support for GitHub Actions today. We could add a WASM workflow for example.

One thing I'd especially like to see, is some kind of ANSI code support, so we can render the Blinkenlights TUI in the browser so that people don't need to run it locally to use it.

**Comment by Vogtinator** [comment]
> One thing I'd especially like to see, is some kind of ANSI code support, so we can render the Blinkenlights TUI in the browser so that people don't need to run it locally to use it.

Getting that to work might not be trivial. emscripten can't do any long-running work on the main thread as it blocks the browser, so waiting for input in native code just does not work. Either the TUI would have to be rewritten to work async or it has to run in a web worker and somehow communicate with the main thread for IO.

**Comment by trungnt2910** [comment]
@jart 

> @trungnt2910 @Vogtinator Would you both be interested in upstreaming your work? I've just added support for GitHub Actions today. We could add a WASM workflow for example.

Before that, there's a problem: After running any binary, the web page must be reloaded in order to run it another time.

See [this commit](https://github.com/trungnt2910/blink/commit/1181a3da830bae754dd79cf12e9874156cbf6bf9) for my workaround. The problem here is, in `memorymalloc.c`, it leaks memory:

```c
void FreeBig(void *p, size_t n) {
  if (!p) return;
#if !defined(__EMSCRIPTEN__)
  unassert(!munmap(p, GetBigSize(n)));
#endif
}
```

If you enable all logging, you can see that it comes from [`Exec`](https://github.com/trungnt2910/blink/blob/5b9f676691b41711a00c437f362eec66879bc76d/blink/blink.c#L115), then `FreeMachine`, `FreeSystem`, and reaches [`FreeBig`](https://github.com/trungnt2910/blink/blob/5b9f676691b41711a00c437f362eec66879bc76d/blink/memorymalloc.c#L203) here. The block of memory that it tries to `munmap` is perfectly valid, but for some reasons the Emscripten runtime always returns `EINVAL`.

**Comment by jart** [comment]
> Getting that to work might not be trivial. emscripten can't do any long-running work on the main thread as it blocks the browser, so waiting for input in native code just does not work. Either the TUI would have to be rewritten to work async or it has to run in a web worker and somehow communicate with the main thread for IO.

Can it just call `setcontext()` or something on entry to `read(0, ...)` to save the CPU state and wait for the JavaScript code to resume it with the result of a keystroke?

**Comment by trungnt2910** [comment]
Hmm, I'm testing a simple Linux "Hello World" x86_64 assembly (the plain old statically ELF file, not any of the special formats here) on the live page, and it always fails:

```
I2023-01-13T21:52:53.783000:blink/throw.c:91:42 SEGMENTATION FAULT AT ADDRESS 0
	 PC 401000 [38;5;155madd [39m[38;5;215m%al[39m,([38;5;215m%rax[39m)
	 AX 0000000000000000  CX 0000000000000000  DX 0000000000000000  BX 0000000000000000
	 SP 00000000f7fffe70  BP 0000000000000000  SI 0000000000000000  DI 0000000000000000
	 R8 0000000000000000  R9 0000000000000000 R10 0000000000000000 R11 0000000000000000
	R12 0000000000000000 R13 0000000000000000 R14 0000000000000000 R15 0000000000000000
	 FS 0000000000000000  GS 0000000000000000 OPS 3                JIT 0               
	executable
	000000000000 000000401000 _start
I2023-01-13T21:52:53.817000:blink/blink.c:66:42 terminating due to signal SIGSEGV
Aborted(native code called abort())
```

**Comment by turbolent** [comment]
Would it be possible to get a WASI or non-Emscripten build, i.e. one that does not require a JS runtime?

**Comment by jart** [comment]
@trungnt2910 could you post a comment with your hello world binary attached? The one we're using is [third_party/cosmo/tinyhello.elf](https://github.com/jart/blink/blob/master/third_party/cosmo/tinyhello.elf)

**Comment by trungnt2910** [comment]
@jart 

> @trungnt2910 could you post a comment with your hello world binary attached? The one we're using is [third_party/cosmo/tinyhello.elf](https://github.com/jart/blink/blob/master/third_party/cosmo/tinyhello.elf)

https://github.com/trungnt2910/HelloElf/blob/master/HelloElf/hello

Here, this one simply writes "Hello World" through a raw `write` syscall and `exit`s immediately.

**Comment by trungnt2910** [comment]
Oops, I sent the wrong binary, the one mentioned above makes an additional `getpid` call.
This one is simpler but still triggers the same issue.

[hello.zip](https://github.com/jart/blink/files/10413852/hello.zip)

**Comment by jart** [comment]
That `hello` binary runs fine under `o//blink/blink` locally. It appears to only be an issue under wasm.

For starters, we really should say `g_high.enabled = false` on the WASM build, so it doesn't emit those ANSI codes.

As for the error, it appears `ThrowSegmentationFault()` is being called when it executes this opcode, which is the very first instruction in your program.

```
  401000:       b8 01 00 00 00          mov    $0x1,%eax
```

It should not be possible for that instruction to segfault. The opcode `b8` runs `OpMovZvqpIvqp` which doesn't have any code paths that would lead it to throw a segfault. It's possible there's something in `GeneralDispatch` that's throwing it. Do you know if it's possible to have WASM print a backtrace of Blink's code?

**Comment by jart** [comment]
Do we know if Blink's JIT is working under WASM? Someone on Hacker News is asking about it. https://news.ycombinator.com/item?id=34368513 We haven't explicitly disabled JIT for Emscripten, and it's able to run programs that have loops, such as third_party/cosmo/loopy.elf until they exit(). So I'd assume it is in fact running JIT'd code under WASM, even though this blog post (https://wingolog.org/archives/2022/08/18/just-in-time-code-generation-within-webassembly) seems to claim that isn't possible because WASM is supposedly a Harvard Architecture. Could it really be the case that blink/jit.c was designed so perfectly that it just works? Or does WASM impose a `W^X` invariant that's causing the JIT to kick the can down the road until a full block can be `mprotect()`'d, at which point it's doomed to fail?

**Comment by trungnt2910** [comment]
> Do you know if it's possible to have WASM print a backtrace of Blink's code?

I can see you have a `PrintBacktrace` function, should I insert this into the `ThrowSegmentationFault` to find out?

**Comment by tkchia** [comment]
Hello @jart,

> It should not be possible for that instruction to segfault. The opcode `b8` runs `OpMovZvqpIvqp` which doesn't have any code paths that would lead it to throw a segfault. It's possible there's something in `GeneralDispatch` that's throwing it.

I see that the error message dump provided by @trungnt2910 lists the instruction as `add %al,(%rax)`, for some reason, rather than the correct `mov $0x1,%eax`.  Thank you!

**Comment by jart** [comment]
> I can see you have a PrintBacktrace function, should I insert this into the ThrowSegmentationFault to find out?

That's unlikely to work. `PrintBacktrace` only works on very few platforms, e.g. asan, ubsan, and darwin's version of libunwind. The Emscripten platform would need to provide an API for doing backtraces that we'd integrate with.

**Comment by tlively** [comment]
> Do we know if Blink's JIT is working under WASM? Someone on Hacker News is asking about it.

Wasm functions and instructions exist entirely separately from Wasm's addressable memory, so there's no way for the JIT to be working unless you're using the workarounds described in Wingo's blog. That said, I have no idea how Blink's JIT works, so maybe it's gracefully falling back to something that still works under Wasm?

**Comment by trungnt2910** [comment]
> Could it really be the case that blink/jit.c was designed so perfectly that it just works?

I doubt it, even .NET's JIT doesn't work under WASM. And your code doesn't seem to have anything related to generating and importing new WASM modules, so the method in that blog post doesn't apply here either.

> Or does WASM impose a W^X invariant that's causing the JIT to kick the can down the road until a full block can be mprotect()'d, at which point it's doomed to fail?

Theoretically, yes, as mentioned on the above comment: "Wasm functions and instructions exist entirely separately from Wasm's addressable memory".

But the curious thing is, Emscripten [stubs](https://github.com/emscripten-core/emscripten/blob/35bcd07b2348c2bc99a3f4cc852d800234fa18a7/system/lib/libc/emscripten_syscall_stubs.c#L208) the call and returns a success status code! So blink is supposed to be tricked into believing that the JIT _succeeded_. I don't know how it magically proceeds after that.

**Comment by jart** [comment]
In that case your PR should update https://github.com/jart/blink/blob/master/blink/builtin.h#L223 to add `!defined(__EMSCRIPTEN__)`.

**Comment by trungnt2910** [comment]
> In that case your PR should update https://github.com/jart/blink/blob/master/blink/builtin.h#L223 to add `!defined(__EMSCRIPTEN__)`.

I don't think so. Now that I see your comment, JIT support instead of being blacklisted is actually _whitelisted_ only for x86_64 and ARM64. WASM, a separate instruction set, is not whitelisted, and JIT is therefore disabled.

**Comment by jart** [comment]
Oh that makes sense! Thank you.

**Comment by fc59283** [comment]
Is it possible to do it with the Zig compiler?
Can Cosmopolitan LibC (or other jart projects) be called from Zig code?

**Comment by Vogtinator** [comment]
> > Getting that to work might not be trivial. emscripten can't do any long-running work on the main thread as it blocks the browser, so waiting for input in native code just does not work. Either the TUI would have to be rewritten to work async or it has to run in a web worker and somehow communicate with the main thread for IO.
> 
> Can it just call `setcontext()` or something on entry to `read(0, ...)` to save the CPU state and wait for the JavaScript code to resume it with the result of a keystroke?

Yes, that's possible: https://emscripten.org/docs/porting/asyncify.html

> That `hello` binary runs fine under `o//blink/blink` locally. It appears to only be an issue under wasm.
> 
> For starters, we really should say `g_high.enabled = false` on the WASM build, so it doesn't emit those ANSI codes.
> 
> As for the error, it appears `ThrowSegmentationFault()` is being called when it executes this opcode, which is the very first instruction in your program.
> 
> ```
>   401000:       b8 01 00 00 00          mov    $0x1,%eax
> ```
> 
> It should not be possible for that instruction to segfault. The opcode `b8` runs `OpMovZvqpIvqp` which doesn't have any code paths that would lead it to throw a segfault. It's possible there's something in `GeneralDispatch` that's throwing it. Do you know if it's possible to have WASM print a backtrace of Blink's code?

I had a quick look. I can reproduce the issue with the `hello` binary here. My guess was that it's an issue with the ELF loader. I enabled logging and it looks like indeed just forgets to load data from the file.

```
I2023-01-13T22:08:28.980000:blink/map.c:50:0 (mem) big created 16m map [0xe0000,0x10e1000)
I2023-01-13T22:08:28.980000:blink/memorymalloc.c:247:0 (thr) new machine thread pid=42 tid=42
I2023-01-13T22:08:28.981000:blink/map.c:50:42 (mem) loader created 8872 map [0x10f0000,0x10f22a8)
I2023-01-13T22:08:28.981000:blink/map.c:50:42 (mem) big created 256k map [0x1110000,0x1150000)
I2023-01-13T22:08:28.982000:blink/loader.c:61:42 (elf) PROGRAM HEADER
I2023-01-13T22:08:28.982000:blink/loader.c:62:42 (elf)   vaddr = 400000
I2023-01-13T22:08:28.982000:blink/loader.c:63:42 (elf)   memsz = e8
I2023-01-13T22:08:28.982000:blink/loader.c:64:42 (elf)   offset = 0
I2023-01-13T22:08:28.982000:blink/loader.c:65:42 (elf)   filesz = e8
I2023-01-13T22:08:28.982000:blink/loader.c:66:42 (elf)   pagesize = 10000
I2023-01-13T22:08:28.982000:blink/loader.c:67:42 (elf)   start = 400000
I2023-01-13T22:08:28.982000:blink/loader.c:68:42 (elf)   end = 410000
I2023-01-13T22:08:28.982000:blink/loader.c:69:42 (elf)   skew = 0
I2023-01-13T22:08:28.982000:blink/loader.c:156:42 (elf) alloc 400000-410000
I2023-01-13T22:08:28.983000:blink/memorymalloc.c:458:42 (mem) reserving virtual [0x400000,0x410000) w/ 64 kb
I2023-01-13T22:08:28.983000:blink/loader.c:164:42 (elf) copy 400000-4000e8 from 0-e8
I2023-01-13T22:08:28.983000:blink/loader.c:61:42 (elf) PROGRAM HEADER
I2023-01-13T22:08:28.983000:blink/loader.c:62:42 (elf)   vaddr = 401000
I2023-01-13T22:08:28.983000:blink/loader.c:63:42 (elf)   memsz = 27
I2023-01-13T22:08:28.983000:blink/loader.c:64:42 (elf)   offset = 1000
I2023-01-13T22:08:28.983000:blink/loader.c:65:42 (elf)   filesz = 27
I2023-01-13T22:08:28.983000:blink/loader.c:66:42 (elf)   pagesize = 10000
I2023-01-13T22:08:28.983000:blink/loader.c:67:42 (elf)   start = 400000
I2023-01-13T22:08:28.983000:blink/loader.c:68:42 (elf)   end = 410000
I2023-01-13T22:08:28.983000:blink/loader.c:69:42 (elf)   skew = 1000
I2023-01-13T22:08:28.983000:blink/loader.c:61:42 (elf) PROGRAM HEADER
I2023-01-13T22:08:28.983000:blink/loader.c:62:42 (elf)   vaddr = 402000
I2023-01-13T22:08:28.983000:blink/loader.c:63:42 (elf)   memsz = e
I2023-01-13T22:08:28.983000:blink/loader.c:64:42 (elf)   offset = 2000
I2023-01-13T22:08:28.983000:blink/loader.c:65:42 (elf)   filesz = e
I2023-01-13T22:08:28.983000:blink/loader.c:66:42 (elf)   pagesize = 10000
I2023-01-13T22:08:28.983000:blink/loader.c:67:42 (elf)   start = 400000
I2023-01-13T22:08:28.983000:blink/loader.c:68:42 (elf)   end = 410000
I2023-01-13T22:08:28.983000:blink/loader.c:69:42 (elf)   skew = 2000
I2023-01-13T22:08:28.983000:blink/memorymalloc.c:458:42 (mem) reserving virtual [0xf7800000,0xf8000000) w/ 8192 kb
I2023-01-13T22:08:28.986000:blink/machine.c:2176:42 (asm) decoding [add %al,(%rax)] at address 401000
I2023-01-13T22:08:28.987000:blink/throw.c:91:42 SEGMENTATION FAULT AT ADDRESS 0
         PC 401000 add %al,(%rax)
```

This is probably due to the 64KiB page size in WASM. FWICT "pages" in wasm only refer to the granularity the memory buffer can be grown at but does not really influence the code running inside. So we can just pretend that it's 4KiB as a workaround, and we get `hello world`!

```diff
diff --git a/blink/map.c b/blink/map.c
index 8fdbc45..0171b6a 100644
--- a/blink/map.c
+++ b/blink/map.c
@@ -28,6 +28,10 @@
 #include "blink/util.h"
 
 long GetSystemPageSize(void) {
+#ifdef __EMSCRIPTEN__
+  return 4096;
+#endif
+
   long z;
   unassert((z = sysconf(_SC_PAGESIZE)) > 0);
   unassert(IS2POW(z));
```

> > I can see you have a PrintBacktrace function, should I insert this into the ThrowSegmentationFault to find out?
> 
> That's unlikely to work. `PrintBacktrace` only works on very few platforms, e.g. asan, ubsan, and darwin's version of libunwind. The Emscripten platform would need to provide an API for doing backtraces that we'd integrate with.

You can `#include <emscripten.h>` and use `EM_ASM(console.trace());` to print a backtrace.

**Comment by duaneking** [comment]
hello.c

```
#include <stdio.h>

int main(){
        printf("Hello World!");
        return 0;
}
```

` gcc -Os -g -static hello.c -o hello -Wl,--hash-style=sysv`

Saddly that gets:

> I2023-01-13T14:26:06.946000:blink/syscall.c:2710:42 missing syscall 0x111
I2023-01-13T14:26:06.946000:blink/syscall.c:2710:42 missing syscall 0x14e
I2023-01-13T14:26:06.947000:blink/syscall.c:1942:42 getrandom() flags not supported yet
malloc(): corrupted top size
I2023-01-13T14:26:06.947000:blink/syscall.c:2414:42 tkill() failed: No such process
I2023-01-13T14:26:06.947000:blink/blink.c:66:42 terminating due to signal SIGABRT
Aborted(native code called abort())


.. but I'm loving what I see so far.

**Comment by jart** [comment]
@duaneking could you attach your hello world binary to this issue tracker?

**Comment by duaneking** [comment]
Sure.

hello.c

```
#include <stdio.h>

int main(){
        printf("Hello World!");
        return 0;
}

```

Compile the a.out

`cc -g -Os -no-pie -fno-pie -static hello.c -U_FORTIFY_SOURCE -fno-stack-protector -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -Wl,-z,common-page-size=65536 -Wl,-z,max-page-size=65536
`

Output from TUI:
```
I2023-01-13T15:02:08.705000:blink/syscall.c:2710:42 missing syscall 0x111
I2023-01-13T15:02:08.707000:blink/syscall.c:2710:42 missing syscall 0x14e
I2023-01-13T15:02:08.708000:blink/syscall.c:1942:42 getrandom() flags not supported yet
malloc(): corrupted top size
I2023-01-13T15:02:08.709000:blink/syscall.c:2414:42 tkill() failed: No such process
I2023-01-13T15:02:08.709000:blink/blink.c:66:42 terminating due to signal SIGABRT
Aborted(native code called abort())

```

Hope that helps.

**Comment by duaneking** [comment]
```
 file a.out
a.out: ELF 64-bit LSB executable, x86-64, version 1 (GNU/Linux), statically linked, BuildID[sha1]=13a011f84eb1bf280499516bd405a19ec4c24c23, for GNU/Linux 3.2.0, with debug_info, not stripped
```

So gcc doesnt use sysv even when requested?

**Comment by duaneking** [comment]
[hello.zip](https://github.com/jart/blink/files/10415975/hello.zip)

**Comment by duaneking** [comment]
file tinyhello.elf
tinyhello.elf: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, no section header

**Comment by jart** [comment]
@duaneking It works for me:

```
$ o//blink/blink ./a.out |& less
I2023-01-13T15:12:30.686823:blink/syscall.c:2859:33689 missing syscall 0x111
I2023-01-13T15:12:30.686904:blink/syscall.c:2859:33689 missing syscall 0x14e
I2023-01-13T15:12:30.687197:blink/syscall.c:1960:33689 getrandom() flags not supported yet
I2023-01-13T15:12:30.688093:blink/xlat.c:524:33689 atf 4096 not supported yet
Hello World!
```

Could you tell us if it works for you if you run it with `-m` to disable linear memory optimization?

```
o//blink/blink -m ./a.out
```

If so, your platform toolchain is probably linking blink in such a way that it's conflicting with the loading of the executable you compiled. Chances are it's being loaded statically to the same addresses. In that case, it'd be really helpful if you could post the top ~50 or so lines of output of:

```
readell -Wa o//blink/blink
```

That will help me confirm my theory, in which case I'll get to finding a way to fix it right away.

**Comment by duaneking** [comment]
I'm not using the command line locally. I'm using the TUI linked above.

**Comment by duaneking** [comment]
you and I are not testing the same things.

I'm using the TUI only. I'm not calling blink locally.

**Comment by duaneking** [comment]
Now pulling blink and running make...

**Comment by duaneking** [comment]
OK so if I compile it locally:

```
 o/blink/blink -m ~/a.out

I2023-01-13T15:41:18.581809:blink/syscall.c:2857:9325 missing syscall 0x111
I2023-01-13T15:41:18.581886:blink/syscall.c:2857:9325 missing syscall 0x14e
I2023-01-13T15:41:18.582101:blink/syscall.c:1960:9325 getrandom() flags not supported yet
I2023-01-13T15:41:18.583009:blink/xlat.c:524:9325 atf 4096 not supported yet
Hello World!
```
It works.

Blink works perfectly locally for me with that binary. But what excited me was the browser part, and that's what I was trying to get working.

**Comment by duaneking** [comment]
Attaching the full readelf output of the binary.

[debug.txt](https://github.com/jart/blink/files/10416092/debug.txt)

**Comment by duaneking** [comment]
I think that's everything I can give you. If you need anything else, just let me know.

**Comment by jart** [comment]
Thank you for the readelf output. What will most likely get you working locally right away is building Blink using our musl-cross-make toolchain. You can do that by running:

```
make -j8 o//x86_64/blink/blink
```

That will build it exactly the same way it builds on my machine. In this case, at HEAD you should see:

```
$ o//x86_64/blink/blink ~/trash/duaneking-specimen.elf
I2023-01-13T17:04:59.550113:blink/syscall.c:2951:36025 missing syscall 0x111
I2023-01-13T17:04:59.550187:blink/syscall.c:2951:36025 missing syscall 0x14e
Hello World!
```

As for these:

```
I2023-01-13T15:41:18.582101:blink/syscall.c:1960:9325 getrandom() flags not supported yet
I2023-01-13T15:41:18.583009:blink/xlat.c:524:9325 atf 4096 not supported yet
```

I've just pushed a few changes adding support for `fstatat(AT_EMPTY_PATH)` and I've suppressed the getrandom() flags warning.

**Comment by trungnt2910** [comment]
> You can `#include <emscripten.h>` and use `EM_ASM(console.trace());` to print a backtrace.

Might it be useful to add this to the existing `PrintBacktrace` function?

**Comment by Vogtinator** [comment]
With `-sASYNCIFY` and some hacks (https://github.com/Vogtinator/blink/commit/12e08b8caf8dd01e4815084754042e9f0f737cfe) I was able to get `blinkenlights.html` to work somewhat:

https://vogtinator.github.io/blink/blinkenlights.html

![Screenshot_20230115_230159](https://user-images.githubusercontent.com/1622084/212569515-1271adaf-45dc-4813-99bb-53bafb656af0.png)

By applying similar hacks to some syscalls (following commits on the master branch), interactive programs such as `vi` and `sh` work in blink.html as well to some extent:

https://vogtinator.github.io/blink/blink.html

![Screenshot_20230115_231948](https://user-images.githubusercontent.com/1622084/212570191-683a9b3e-6071-47ae-bda2-2b48d5fd2d69.png)

![Screenshot_20230115_232027](https://user-images.githubusercontent.com/1622084/212570213-5b9f3592-fe05-4991-b7cf-c86b30b979f2.png)

**Comment by duaneking** [comment]
Just pulled the most recent changes and updated locally;

I'm loving that the only error output I get is the 0x14e syscall.

I think on linux that *might* be the compat_sendmmsg syscall?

Either way, Thanks for the fast update. 

I reused the old client binaries with the new updated blink build, so nothing in terms of test input has changed outside of the upgrade.

```
$ o/blink/blink -m ~/a.out
I2023-01-15T14:23:39.729686:blink/syscall.c:3265:28627 missing syscall 0x14e
Hello World!
```

**Comment by jart** [comment]
Wow I'm really impressed to see Vi running in Blink in the browser. I'd love to be able to tweet that, once we get it up on https://jart.github.io/blink/blink.html

Since you're using BusyBox, I'd like to give you something even better. Here's what I call the "Blink Software Pack" which contains Actually Portable Executables from the Cosmopolitan project that are known to work well under Blink (because Blink was designed primarily with Cosmopolitan in mind). The listing is:

- qjs.com: Fabrice Bellard's QuickJS interpreter
- sqlite3.com: Standard SQLite command
- redbean.com: HTTP/HTTPS server with Lua and SQLite baked-in
- lua.com: Standard Lua interpreter with Redbean's UNIX module
- printimage.com: Renders GIF/JPG/PNG in terminal as XTERM UNICODE block characters
- finger.com: Finger protocol client, e.g. `./finger.com new_york@graph.no`
- curl.com: Lightweight replacement for the curl HTTP client command
- unbourne.com: Fork of the Almquist shell with Bestline, a BSD-licensed GNU Readline replacement
- uname.com: Invokes the uname(2) system call
- ttyinfo.com: Interactive troubleshooting tool for ANSI escape codes
- sysinfo.com: Invokes the sysinfo(2) system call
- kilo.com: Antirez's famous 1000 LOC text editor
- hangman.com: The famous text-based game from UNIX Seventh Edition
- tidy.com: The HTML5 Tidy command
- ctags.com: Exuberant Ctags
- make.com: GNU Make w/ Justine's Landlock LSM security extensions
- python.com: Cosmopolitan's Python 3.6 interpreter in a single file
- awk.com: The One True Awk
- sha256sum.com: Replacement for the sha256sum command that goes 2x faster on modern x86
- mkdeps.com: Fast tool for generating Makefile code based on header file dependencies
- gzip.com: Compresses files using Zlib's DEFLATE implementation
- echo.com: Tool that writes a line of text to standard output
- rm.com: Tool for removing files
- mkdir.com: Tool for making directories
- pwd.com: Tool for printing current directory
- touch.com: Tool for updating timestamps on files
- cp.com: Tool for copying files
- dd.com: Tool for transferring blocks of data between files
- printf.com: Tool that formats a line of text to standard output
- chmod.com: Tool for changing file permissions

You can download the software pack here: 

https://justine.lol/blinkenlights/blink-software-pack.tar.gz

One issue you may encounter with WASM in the browser, is many Cosmopolitan binaries assume they have the ability to `open(argv[0])` so the WASM runtime should provide the executable access to read its own content. Otherwise programs like python.com won't be able to read the ZIP files contained within the executable.

**Comment by unicomp21** [comment]
Wow, you guys are amazing!

**Comment by duaneking** [comment]
I suggest that a sub-project be made to allow such software packs but as a more formal release, perhaps something such as blinkOS? ;)

**Comment by unicomp21** [comment]
I can't wait to see what happens w/ blinkOS, should be interesting!

**Comment by unicomp21** [comment]
Not sure if it's applicable here, but I know the leaningtech guys are pretty excited about having tail calls for webvm.  Looks like they will land soon in both chrome and safari.

https://github.com/WebAssembly/tail-call/issues/15#issuecomment-1380699814

**Comment by jart** [comment]
> I'm loving that the only error output I get is the 0x14e syscall. I think on linux that might be the compat_sendmmsg syscall?

That's `rseq` which is a deeply-Linux threading-related system call introduced a couple years ago that every Glibc binary seems to need. It doesn't have a man page aside this https://lwn.net/Articles/774098/ We should examine Glibc to confirm it has a fallback path for us returning ENOSYS, in which case we can suppress the warning.

https://github.com/jart/cosmopolitan/blob/be3e109309d7adc0824b8c22cf4fca3aa6433baa/libc/sysv/syscalls.sh#L370

**Comment by duaneking** [comment]
> > I'm loving that the only error output I get is the 0x14e syscall. I think on linux that might be the compat_sendmmsg syscall?
> 
> That's `rseq` which is a deeply-Linux threading-related system call introduced a couple years ago that every Glibc binary seems to need. It doesn't have a man page aside this https://lwn.net/Articles/774098/ We should examine Glibc to confirm it has a fallback path for us returning ENOSYS, in which case we can suppress the warning.
> 
> https://github.com/jart/cosmopolitan/blob/be3e109309d7adc0824b8c22cf4fca3aa6433baa/libc/sysv/syscalls.sh#L370

Yeah I was thinking an older version; https://thevivekpandey.github.io/posts/2017-09-25-linux-system-calls.html

Good catch.

**Comment by duaneking** [comment]
Linux rseq link for Reference: https://github.com/torvalds/linux/blob/master/kernel/rseq.c

I'm idling through code to see what I can find.

So it seems https://www.gnu.org/software/libc/manual/html_node/Restartable-Sequences.html confirms "The GNU C Library sets the cpu_id field to RSEQ_CPU_ID_REGISTRATION_FAILED if registration failed or was explicitly disabled." so I'm looking and https://github.com/torvalds/linux/blob/master/include/uapi/linux/rseq.h says thats defaulted to -2 currently.

This looks interesting:

https://sourceware.org/git/?p=glibc.git;a=blob;f=nptl/pthread_create.c;h=4608fd9068a56e80310f93fd229c96e344257f91;hb=e3e589829d16af9f7e73c7b70f74f3c5d5003e45

**Comment by duaneking** [comment]
For future reference: https://sourceware.org/git/?p=glibc.git&a=search&h=HEAD&st=commit&s=rseq

**Comment by duaneking** [comment]
Ok found the rseq glibc code:

https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/sys/rseq.h;h=8533782cf4acf945a270abcbaa100bc806bf6cd8;hb=6c33b018438ee799c29486f21d43d8100bdbd597

**Comment by duaneking** [comment]
It looks like rseq support is mandated/required on the versions of linux that support it.

https://sourceware.org/git/?p=glibc.git;a=blob;f=nptl/pthread_create.c;hb=569cfcc6bf35c28112ca8d7112e9eb4a22bed5b8#l376

And that shows that not supporting rseq is a fatal glibc panic/error.

**Comment by unicomp21** [comment]
"if" we'll soon have tail calls, which means emscripten will then support co_await and co_yield?  Along with the compiler level optimizations they both provide?  Does the need for emscripten "asyncify" go away?

https://github.com/jart/blink/issues/8#issuecomment-1383286674

**Comment by unicomp21** [comment]
actually, maybe only co_await requires tail calls?  and co_yield already works?  @tlively

**Comment by Niek** [comment]
Running `busybox` and `vim` is very impressive, a minimal Alpine would be even cooler (similar to https://bellard.org/jslinux/ / TinyEMU). Or is that out of scope of this project?

**Comment by unicomp21** [comment]
There are images here too ...
https://github.com/copy/v86

**Comment by beriberikix** [comment]
> Wow I'm really impressed to see Vi running in Blink in the browser. I'd love to be able to tweet that, once we get it up on https://jart.github.io/blink/blink.html
> 
> Since you're using BusyBox, I'd like to give you something even better. Here's what I call the "Blink Software Pack" which contains Actually Portable Executables from the Cosmopolitan project that are known to work well under Blink (because Blink was designed primarily with Cosmopolitan in mind). The listing is:
> 
> * qjs.com: Fabrice Bellard's QuickJS interpreter
> * sqlite3.com: Standard SQLite command
> * redbean.com: HTTP/HTTPS server with Lua and SQLite baked-in
> * lua.com: Standard Lua interpreter with Redbean's UNIX module
> * printimage.com: Renders GIF/JPG/PNG in terminal as XTERM UNICODE block characters
> * finger.com: Finger protocol client, e.g. `./finger.com new_york@graph.no`
> * curl.com: Lightweight replacement for the curl HTTP client command
> * unbourne.com: Fork of the Almquist shell with Bestline, a BSD-licensed GNU Readline replacement
> * uname.com: Invokes the uname(2) system call
> * ttyinfo.com: Interactive troubleshooting tool for ANSI escape codes
> * sysinfo.com: Invokes the sysinfo(2) system call
> * kilo.com: Antirez's famous 1000 LOC text editor
> * hangman.com: The famous text-based game from UNIX Seventh Edition
> * tidy.com: The HTML5 Tidy command
> * ctags.com: Exuberant Ctags
> * make.com: GNU Make w/ Justine's Landlock LSM security extensions
> * python.com: Cosmopolitan's Python 3.6 interpreter in a single file
> * awk.com: The One True Awk
> * sha256sum.com: Replacement for the sha256sum command that goes 2x faster on modern x86
> * mkdeps.com: Fast tool for generating Makefile code based on header file dependencies
> * gzip.com: Compresses files using Zlib's DEFLATE implementation
> * echo.com: Tool that writes a line of text to standard output
> * rm.com: Tool for removing files
> * mkdir.com: Tool for making directories
> * pwd.com: Tool for printing current directory
> * touch.com: Tool for updating timestamps on files
> * cp.com: Tool for copying files
> * dd.com: Tool for transferring blocks of data between files
> * printf.com: Tool that formats a line of text to standard output
> * chmod.com: Tool for changing file permissions
> 
> You can download the software pack here:
> 
> https://justine.lol/blinkenlights/blink-software-pack.tar.gz
> 
> One issue you may encounter with WASM in the browser, is many Cosmopolitan binaries assume they have the ability to `open(argv[0])` so the WASM runtime should provide the executable access to read its own content. Otherwise programs like python.com won't be able to read the ZIP files contained within the executable.

Curious if anyone has tried these lately. They all are erroring for me, albeit is slightly different ways.

**Comment by trungnt2910** [comment]
Due to certain bugs in the CI process, the WASM preview has not been updated for a long time. You might have better luck trying a local build.

**Comment by RNRetailer** [comment]
Hi.

I'm trying to run a Linux binary in the browser.

It's written in Golang.

I can't compile for WASM because it has so many Linux dependencies.

I tried to run it here: https://trungnt2910.com/blink/blink.html

However, I got this error:
"Uncaught RuntimeError: Aborted(Cannot enlarge memory arrays to size 1073811456 bytes (OOM). Either (1) compile with -sINITIAL_MEMORY=X with X higher than the current value 1073741824, (2) compile with -sALLOW_MEMORY_GROWTH which allows increasing the size at runtime, or (3) if you want malloc to return NULL (0) instead of this abort, compile with -sABORTING_MALLOC=0)"

Does this mean that emscripten must be compiled with -sALLOW_MEMORY_GROWTH?

I can pay if someone wants to help.

Thanks

**Comment by jart** [comment]
@RNRetailer If you can help fund further development on WASM integration, then please contact me by email.

**Comment by trungnt2910** [comment]
> I can pay if someone wants to help.

I'm one of the authors of the WASM port and the owner of the page you mentioned. The page on my fork is a bit outdated compared to the [official one](https://jart.github.io/blink/blink.html), which itself has also been abandoned for quite a long time. 

If you need my help updating the WASM version and fixing specific bugs, please reach out to me at `me` [at] `trungnt2910.com` or through my [Discord](https://reality.trungnt2910.com/discord).

**Comment by Vogtinator** [comment]
> Does this mean that emscripten must be compiled with -sALLOW_MEMORY_GROWTH?

No, just build blink with one of the options.

However:

> I can't compile for WASM because it has so many Linux dependencies.

Depending on the dependencies you might have to spend more time getting those to work with blink wasm. With the additional layer in between you'll also significantly lose performance.

**Comment by baudaux** [comment]
Hi, I have successfully compiled and integrated this great project in exaequOS, the OS I am building and that fully runs in a Web Browser. Then now, it is possible to import elf file in the browser and execute it. In the future graphical apps will be supported (OpenGL, Wayland). You can type  in a terminal: blink /usr/bin/x86_64/nano

**Comment by xlab** [comment]
<img width="983" height="146" alt="Image" src="https://github.com/user-attachments/assets/7fcebc31-09f3-469e-b2b1-503228fec3ee" />

Uh, needs work

To compare, RISCV64 via TinyEmu:

```
Math.sqrt: 10970.70 ms for 1000000 iterations
```

RISCV64 on QEMU/docker (outside of browser):

```
Math.sqrt: 1103.01 ms for 1000000 iterations
```
